### DISCUSSION TITLE
How to get value from PropertyGrid cell developed from C#

### DISCUSSION THREAD

**Discussion by zosun** [pywinauto/pywinauto#1230] [Q&A]
I have one C# app which contains PropertyGrid control. How can I get the value from some cells in the PropertyGrid control with Pywinauto? thanks

**Comment by yihleego** [comment]
Hi @zosun,

I think what you need is **Inspect**, it is a Windows-based tool that enables you select any UI element and view the element's accessibility data. It has several main parts:

- Title bar. Displays the Inspect window handle (HWND).
- Menu bar. Provides access to Inspect functionality.
- Toolbar. Provides access to Inspect functionality.
- Tree view. Presents the hierarchical structure of UI elements as a tree-view control that you can use to navigate among the elements.
- Data view. Displays all exposed accessibility properties for the selected UI element.

![image](https://user-images.githubusercontent.com/21337386/183144040-536110c9-a8a2-4f70-876b-811b2807bd53.png)

---

**Inspect** is installed as part of the overall set of tools in the Windows Software Development Kit (SDK), it is not distributed as a separate download. The Windows SDK includes all of the accessibility-related tools documented in this section.

[Download the Windows SDK](https://developer.microsoft.com/en-us/windows/downloads/windows-10-sdk/)

`Inspect.exe` is located in the \bin\<version>\<platform> folder of the SDK installation path (you don't typically have to run as administrator).



You can also download `Inspect.exe` from [this repository](https://github.com/yihleego/Windows-Kits/tree/master/Inspect).


See also: https://docs.microsoft.com/en-us/windows/win32/winauto/inspect-objects

**Reply by zosun** [reply]
Hi Leego,

thanks for your hints. inspect is replaced by new tool (Accessibility Insights for Window) right now. 

I checked the UI element with it.  It said that the cells in PropertyGrid is custom type.

I also checked  the Microsoft UI Automation document which says that PropertyGrid is not supported by them.

Thanks/Zongjun

**Comment by vasily-v-ryabov** [comment]
If it's WPF app, we have experimental `backend="wpf"` implementation here: https://github.com/pywinauto/pywinauto/pull/1210 (not merged yet, but should be fully functional). Any feedback is welcome!

**Reply by eltimen** [reply]
Hello @zosun. Does your application use a PropertyGrid control from Windows Forms or WPF (DevExpress, Syncfusion, Telerik, ...)?

If it is WPF control, I would be interested in testing the new WPF backend that works via DLL injection and uses the visual/logical tree to find GUI controls and interacts with their corresponding .NET objects (via getting/setting .NET properties, method calls, etc.). 
This approach is probably the only way to automate thirdparty GUI controls that do not support MS UI Automation.

**Reply by zosun** [reply]
Hi @eltimen and @vasily-v-ryabov. the application was developed by other vendors.  Per my understanding, they may use PropertyGrid control from Windows Forms.  From microsoft website, there is no PropertyGrid in WPF controls set.

Does backend=wpf work for Windows Form application?

Thanks

**Reply by eltimen** [reply]
Unfortunately, this backend doesn't support Windows Forms applications.

**Reply by vasily-v-ryabov** [reply]
It's technically possible to create a new backend for WinForms custom controls, but it will take pretty much time. If the application is publicly available, can you please share it?

**Reply by zosun** [reply]
[WinFormsApp.zip](https://github.com/pywinauto/pywinauto/files/9312949/WinFormsApp.zip)
Hi @vasily-v-ryabov  the application is not publicly available, but I created one which utilizes the same PropertyGrid Control. Please check executable contained in the zip file. thanks

**Reply by vasily-v-ryabov** [reply]
@zosun thanks! This is very helpful. Unfortunately we had no chance to look into it. We will keep it for the future.

**Reply by sachinpatole** [reply]
@vasily-v-ryabov I want to automate a windows application ([UA Expert](unified-automation.com/downloads/opc-ua-clients.html)) developed in PyQT-4; however when I inspect UI object using Inspect.exe, all UI objects are NOT visible in Inspect.exe tool. I tried with both the backends.

Is there any setting which needs to be done for exploring all UI object of application developed in PyQT.

**Reply by vasily-v-ryabov** [reply]
@sachinpatole try to set env variable `QT_ACCESSIBILITY=1` or `QT_USE_NATIVE_WINDOWS=1` like it was mentioned here https://stackoverflow.com/a/26465446/3648361
